src/lib/comms/sol-coap.c

```diff
@@ -678,11 +684,49 @@ timeout_expired(struct sol_coap_server *server, struct outgoing *outgoing)
     return expired;
 }
 
+static struct sol_buffer *prepare_buffer(struct outgoing *outgoing)
+{
+    static struct sol_buffer buffer;
+    struct sol_coap_packet *header = outgoing->header;
+    struct sol_coap_packet *payload = outgoing->pkt;
+    uint16_t new_size, new_offset, old_offset;
+    uint8_t new_tkl, old_tkl;
+    int r;
+
+    if (!header) {
+        buffer = SOL_BUFFER_INIT_CONST(payload->buf.data, payload->buf.used);
+        return &buffer;
+    }
+
+    (void)sol_coap_header_get_token(payload, &old_tkl);
+    (void)sol_coap_header_get_token(header, &new_tkl);
+
+    new_size = payload->buf.used + (new_tkl - old_tkl);
+
+    sol_buffer_init(&buffer);
+
+    r = sol_buffer_ensure(&buffer, new_size);
+    SOL_INT_CHECK(r, < 0, NULL);
+
+    new_offset = sizeof(struct coap_header) + new_tkl;
+    old_offset = sizeof(struct coap_header) + old_tkl;
+
+    r = sol_buffer_insert_bytes(&buffer, 0, header->buf.data, new_offset);
+    SOL_INT_CHECK(r, < 0, NULL);
+
+    r = sol_buffer_insert_bytes(&buffer, new_offset,
+        (uint8_t *) payload->buf.data + old_offset, payload->buf.used - old_offset);
```

Comment:
`sol_buffer_get_at(&payload->buf, old_offset)`

check the `sol-buffer.h`, you need a refresh man!